### PR TITLE
TDL-20059: Implement currently_syncing

### PR DESCRIPTION
**Pull request by harshpatel4crest**
# Description of change
TDL-20059: Implement currently_syncing
- Added currently syncing functionality

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_quickbooks/sync.py [hunk 1]
```diff
@@ -24,6 +24,9 @@ def do_sync(client, config, state, catalog):
         )
 
         LOGGER.info("Syncing stream: %s", stream_id)
+        # Set currently syncing stream
+        state = singer.set_currently_syncing(state, stream_id)
```
Comment (dbshah1212) [inline tap_quickbooks/sync.py:28]: Where do we handle the use of this?

**Review by savan-chovatiya — APPROVED** [review]

**Review by rdeshmukh15 — COMMENTED** [review]
Can you please add the integration test case scenario for the interrupted sync which will validate the changes done in sync.py?

**Comment by harshpatel4crest** [comment]
> Can you please add the integration test case scenario for the interrupted sync which will validate the changes done in sync.py?

Added tap-tester test for interrupted sync.

File: tests/test_quickbooks_all_fields.py [hunk 2]
```diff
@@ -246,7 +246,9 @@ class TestQuickbooksAllFields(TestQuickbooksBase):
     }
 
     def name(self):
-        return "tap_tester_quickbooks_combined_test"
+        # Quickbooks uses the token chaining to get the existing token which requires all tests to have same name
+        # So do not overwrite the test name below
```
Comment (rdeshmukh15) [inline tests/test_quickbooks_all_fields.py:250]: make the same change here as well, use docstring

https://peps.python.org/pep-0257/

File: tests/test_quickbooks_automatic_fields.py [hunk 3]
```diff
@@ -9,7 +9,9 @@
 
 class TestQuickbooksAutomaticFields(TestQuickbooksBase):
     def name(self):
-        return "tap_tester_quickbooks_combined_test"
+        # Quickbooks uses the token chaining to get the existing token which requires all tests to have same name
+        # So do not overwrite the test name below
```
Comment (rdeshmukh15) [inline tests/test_quickbooks_automatic_fields.py:13]: Use docstring here as well

File: tests/test_quickbooks_discovery.py [hunk 4]
```diff
@@ -7,7 +7,9 @@
 
 class TestQuickbooksDiscovery(TestQuickbooksBase):
     def name(self):
-        return "tap_tester_quickbooks_combined_test"
+        # Quickbooks uses the token chaining to get the existing token which requires all tests to have same name
+        # So do not overwrite the test name below
```
Comment (rdeshmukh15) [inline tests/test_quickbooks_discovery.py:11]: Use docstring like this :
"""
Quickbooks uses the token chaining to get the existing token which requires all tests to have same name
So do not overwrite the test name below
"""

File: tests/test_quickbooks_interrupted_sync.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
+from datetime import datetime as dt
+from tap_tester import runner, connections, menagerie
+from base import TestQuickbooksBase
+
+class TestQuickbooksInterruptedSyncTest(TestQuickbooksBase):
+
+    def assertIsDateFormat(self, value, str_format):
+        """
+            Assertion Method that verifies a string value is a formatted datetime with
```
Comment (rdeshmukh15) [inline tests/test_quickbooks_interrupted_sync.py:9]: Indentation of line 9 and 10 should be same as line no. 8

File: tests/test_quickbooks_interrupted_sync.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            the specified format.
+        """
+        try:
+            dt.strptime(value, str_format)
+        except ValueError as err:
+            raise AssertionError(f"Value: {value} does not conform to expected format: {str_format}") from err
+
+    def name(self):
+        # Quickbooks uses the token chaining to get the existing token which requires all tests to have same name
+        # So do not overwrite the test name below
```
Comment (rdeshmukh15) [inline tests/test_quickbooks_interrupted_sync.py:19]: use docstring here as well

File: tests/test_quickbooks_pagination.py [hunk 7]
```diff
@@ -8,7 +8,9 @@
 
 class TestQuickbooksPagination(TestQuickbooksBase):
     def name(self):
-        return "tap_tester_quickbooks_combined_test"
+        # Quickbooks uses the token chaining to get the existing token which requires all tests to have same name
+        # So do not overwrite the test name below
+        return super().name()
```
Comment (rdeshmukh15) [inline tests/test_quickbooks_pagination.py:13]: use docstring here as well

File: tests/test_quickbooks_start_date.py [hunk 8]
```diff
@@ -6,7 +6,9 @@
 
 class TestQuickbooksStartDate(TestQuickbooksBase):
     def name(self):
-        return "tap_tester_quickbooks_combined_test"
+        # Quickbooks uses the token chaining to get the existing token which requires all tests to have same name
+        # So do not overwrite the test name below
```
Comment (rdeshmukh15) [inline tests/test_quickbooks_start_date.py:10]: use docstring here as well

File: tests/test_quickbooks_sync_all.py [hunk 9]
```diff
@@ -7,7 +7,9 @@
 
 class TestQuickbooksSyncAll(TestQuickbooksBase):
     def name(self):
-        return "tap_tester_quickbooks_combined_test"
+        # Quickbooks uses the token chaining to get the existing token which requires all tests to have same name
+        # So do not overwrite the test name below
```
Comment (rdeshmukh15) [inline tests/test_quickbooks_sync_all.py:11]: use docstring here as well

Comment (harshpatel4crest) [reply tests/test_quickbooks_sync_all.py]: Added docstring for the function.

**Comment by harshpatel4crest** [comment]
@rdeshmukh15 Added docstring for the functions.

File: tests/test_quickbooks_interrupted_sync.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+                    if stream == state['currently_syncing']:
+
+                        # Check if the interrupted stream has a bookmark written for it
+                        if state["bookmarks"].get(stream):
+                            interrupted_stream_bookmark = self.convert_state_to_utc(state['bookmarks'].get(stream).get('LastUpdatedTime'))
+                            interrupted_bookmark_datetime = dt.strptime(interrupted_stream_bookmark, "%Y-%m-%dT%H:%M:%SZ")
+                        else:
+                            # Assign the start date to the interrupted stream
+                            interrupted_bookmark_datetime = start_date_datetime
```
Comment (kspeer825) [inline tests/test_quickbooks_interrupted_sync.py:160]: Not sure i understand the purpose of this logic. This block will only only execute against the interrupted `bill_payments` stream, so the reference point should be the state that was set, not the start date.

File: tests/test_quickbooks_interrupted_sync.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+
+                        # - Verify resuming sync only replicates records with replication key values greater or
+                        #       equal to the state for streams that were replicated during the interrupted sync.
+                        # - Verify the interrupted sync replicates the expected record set all interrupted records are in full records
+                        for record in interrupted_records:
```
Comment (kspeer825) [inline tests/test_quickbooks_interrupted_sync.py:165]: Can you condense the assertions in this test? There is a lot of unnecessary conditional logic and repeated assertions. It decreases readability.

**Review by kspeer825 — CHANGES_REQUESTED** [review]

File: tests/base.py [hunk 12]
```diff
@@ -257,3 +266,12 @@ def dt_to_ts(self, dtime):
 
     def is_report_stream(self, stream):
         return stream in ["profit_loss_report"]
+
+    def convert_state_to_utc(self, date_str):
+        """
+        Convert a saved bookmark value of the form '2020-08-25T13:17:36-07:00' to a string
+        formatted utc datetime, in order to compare against the json formatted datetime values
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/base.py:278]: Yes, this is the function from the existing test case([here](https://github.com/singer-io/tap-quickbooks/blob/master/tests/test_quickbooks_bookmarks.py#L21)). As we want to use it in the interrupted sync test case, we have moved it into the base test file.

File: tests/test_quickbooks_automatic_fields.py [hunk 13]
```diff
@@ -9,7 +9,11 @@
 
 class TestQuickbooksAutomaticFields(TestQuickbooksBase):
     def name(self):
-        return "tap_tester_quickbooks_combined_test"
+        """
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/test_quickbooks_automatic_fields.py]: Removed `name` from all the test cases.

File: tests/test_quickbooks_interrupted_sync.py [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+        if original:
+            return {
+                'start_date' : '2016-06-02T00:00:00Z',
+                'sandbox': 'true'
+            }
+        else:
+            return {
+                'start_date' : self.start_date,
+                'sandbox': 'true'
+            }
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/test_quickbooks_interrupted_sync.py]: Moved the `get_properties` into the base test file.

File: tests/test_quickbooks_interrupted_sync.py [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+    def test_run(self):
+        """
+        Scenario: A sync job is interrupted. The state is saved with `currently_syncing`.
+                  The next sync job kicks off and the tap picks back up on that `currently_syncing` stream.
+        Expected State Structure:
+            {
+                "currently_syncing": "stream-name",
+                "bookmarks": {
+                    "stream-name-1": {"LastUpdatedTime": "2021-06-10T11:46:26-07:00"},
+                    "stream-name-2": {"LastUpdatedTime": "2021-06-10T11:46:26-07:00"}
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tests/test_quickbooks_interrupted_sync.py:26]: Updated the hardcoded state file as the state file format from the Tap.

File: tests/test_quickbooks_interrupted_sync.py (hunk 10 above)
Comment (harshpatel4crest) [reply tests/test_quickbooks_interrupted_sync.py]: Updated the assertions for the interrupted stream to use the start date as the reference point.

File: tests/test_quickbooks_interrupted_sync.py (hunk 11 above)
Comment (harshpatel4crest) [reply tests/test_quickbooks_interrupted_sync.py:141]: We condensed some assertions for non-interrupted streams. But for the interrupted streams, we verify that we synced all the records >= start date and verify the count of the records based on the 1st sync.

File: tests/test_quickbooks_interrupted_sync.py [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+                        # - Verify resuming sync only replicates records with replication key values greater or equal to
+                        #       the state for streams that were replicated during the interrupted sync.
+                        # - Verify resuming sync replicates all records that were found in the full sync (non-interupted)
+                        for record in interrupted_records:
+                            rec_time = dt.strptime(record.get('MetaData').get('LastUpdatedTime'), "%Y-%m-%dT%H:%M:%S.%fZ")
+                            self.assertGreaterEqual(rec_time, synced_stream_datetime)
+
+                            self.assertIn(record, full_records, msg='Unexpected record replicated in resuming sync.')
+
+                elif expected_replication_method == self.FULL_TABLE:
```
Comment (kspeer825) [inline tests/test_quickbooks_interrupted_sync.py:177]: There are no full table streams under test here. There don't appear to be any full table streams in the tap based on the expectations in base.py

**Review by kspeer825 — CHANGES_REQUESTED** [review]

File: tests/test_quickbooks_interrupted_sync.py (hunk 16 above)
Comment (harshpatel4crest) [reply tests/test_quickbooks_interrupted_sync.py]: Updated the code and removed conditions for checking the replication method for expected streams.

**Review by rdeshmukh15 — COMMENTED** [review]
Use dosctring at line no 6 in test_quickbooks_all_fields.py

**Comment by rdeshmukh15** [comment]
No need of line no 27 
``` python
    FULL = "FULL_TABLE"
```
As all streams are INCREMENTAL and the above line is not used anywhere.

**Comment by rdeshmukh15** [comment]
I think, no need of writing one function like this :
In file test_quickbooks_automatic_fields.py 
```python
    def expected_streams(self):
        """
        All streams are under test with the exception of 'budgets' which
        has a workaround to skip the invalid assertion. See if block in
        test_run for details
        """
        return self.expected_check_streams()
```
instead of this, you can directly call the function and store the value in the variable like below :
``` python 
expected_streams = self.expected_check_streams()
```

**Comment by rdeshmukh15** [comment]
In file test_quickbooks_bookmarks.py instead of this :
``` python
    def expected_streams(self):
        return self.expected_check_streams().difference({
            'budgets'
        })
```
write this to reduce no of lines of code :
``` python
expected_streams =  self.expected_check_streams().difference({'budgets'})
```
And use the variable expected_streams wherever required

**Comment by rdeshmukh15** [comment]
In file test_quickbooks_bookmarks.py, 
instead of checking the condition `self.is_report_stream(stream)` multiple times.
You can check it once and can do something like this :
```python
                if self.is_report_stream(stream) :
                       top_level_replication_key ='ReportDate'
                       simulated_bookmark_value -= 2592000
                 for message in second_sync_messages:
                        replication_key_value = message.get('data').get(top_level_replication_key).get(sub_level_replication_key)
                    self.assertGreaterEqual(self.dt_to_ts(replication_key_value), simulated_bookmark_value,
                                            msg="Second sync records do not repect the previous bookmark.")
```
Please do the same kind of changes for other blocks of for as well

**Comment by rdeshmukh15** [comment]
Remove below block from test_quickbooks_discovery.py :
```python
                else:
                    self.assertTrue(actual_replication_method == self.FULL,
                                    msg="Expected FULL replication "
                                        "since there is no replication key")
```
As there are no full table streams present for this tap

**Comment by harshpatel4crest** [comment]
@rdeshmukh15 As discussed, updated the tap-tester test comments as part of PR #61 as the PR is dedicated to the tap-tester tests.

**Comment by rdeshmukh15** [comment]
In test_quickbooks_interrupted_sync.py,
full_record_count and record_count_by_stream_full_sync variables are calculated but are not used anywhere.
Can you please remove both of them?

**Comment by harshpatel4crest** [comment]
> In test_quickbooks_interrupted_sync.py, full_record_count and record_count_by_stream_full_sync variables are calculated but are not used anywhere. Can you please remove both of them?

Removed unused variables from the interrupted sync test case.

**Review by rdeshmukh15 — APPROVED** [review]

**Pull request merged**